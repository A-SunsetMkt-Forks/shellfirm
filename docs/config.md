# Config

When you install `shellfirm` the first time it creates a new config file in the home directory in the path: `~/.shellfirm/config.yaml`.
The main configuration file manage in [config.yaml](../src/config.yaml). 

You can always change your config file content and the `shellfirm` will never change it back. 
[read here how to add and test new command](./contributing.md#test-new-command)


# Config
| Parameter | Description | Values |
| --- | --- | --- |
| `challenge` | The way that you want to solve the challenge when risky command detected | `Math`, `Enter`, `Yes` |
| `includes` | List of group checks. | `list` |
| `checks[].is` | The value of the check | `String` |
| `checks[].method` | How to make the check | `Contains`, `Regex`, `StartWith` |
| `checks[].enable` | Enable disable the check | `true`, `false` |
| `checks[].description` | Prompt description when risky command directed | `String` |
| `checks[].from` | Group name | `String` |


# Update config file

Adding new groups:
```bash
$ shellfirm config update --check-group {group} {group}
```

Remove groups
```bash
$ shellfirm config update --check-group {group} {group} --remove
```

Reset 
```bash
$ shellfirm config reset
```