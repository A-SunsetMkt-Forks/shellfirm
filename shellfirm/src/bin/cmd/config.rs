use anyhow::{anyhow, Result};
use clap::{App, Arg, ArgMatches, Command};
use shellfirm::{dialog, Challenge, Config};

const ALL_GROUP_CHECKS: &[&str] = &include!(concat!(env!("OUT_DIR"), "/all_the_files.rs"));

pub fn command() -> Command<'static> {
    Command::new("config")
        .about("Manage app config")
        .subcommand(
            App::new("update-groups")
                .about("enable check group")
                .arg(Arg::new("check-group").help("Check group")),
        )
        .subcommand(App::new("reset").about("Reset configuration"))
        .subcommand(
            App::new("challenge").about("Reset configuration").arg(
                Arg::new("challenge")
                    .possible_values(&["Math", "Enter", "Yes"])
                    .required(true)
                    .takes_value(true),
            ),
        )
}

pub fn run(matches: &ArgMatches, config: &Config) -> Result<shellfirm::CmdExit> {
    match matches.subcommand() {
        None => Err(anyhow!("command not found")),
        Some(tup) => match tup {
            ("update-groups", _subcommand_matches) => run_update_groups(config),
            ("reset", _subcommand_matches) => run_reset(config),
            ("challenge", subcommand_matches) => run_challenge(subcommand_matches, config),
            _ => unreachable!(),
        },
    }
}

pub fn run_update_groups(config: &Config) -> Result<shellfirm::CmdExit> {
    let all_groups = ALL_GROUP_CHECKS.iter().map(|f| f.to_string()).collect();
    let settings = config.get_settings_from_file()?;

    let check_groups = dialog::multi_choice(
        "select checks",
        all_groups,
        settings.get_active_groups().to_vec(),
        100,
    )?;

    let res: Vec<String> = check_groups
        .iter()
        .map(std::string::ToString::to_string)
        .collect();

    match config.update_check_groups(res) {
        Ok(()) => Ok(shellfirm::CmdExit {
            code: exitcode::OK,
            message: None,
        }),
        Err(e) => Ok(shellfirm::CmdExit {
            code: exitcode::CONFIG,
            message: Some(format!("Could not update checks group. err: {}", e)),
        }),
    }
}

pub fn run_reset(settings: &Config) -> Result<shellfirm::CmdExit> {
    settings.reset_config(None)?;

    Ok(shellfirm::CmdExit {
        code: exitcode::OK,
        message: Some("shellfirm configuration reset successfully".to_string()),
    })
}

pub fn run_challenge(matches: &ArgMatches, settings: &Config) -> Result<shellfirm::CmdExit> {
    let challenge = match matches.value_of("challenge").unwrap() {
        "Enter" => Challenge::Enter,
        "Yes" => Challenge::Yes,
        _ => Challenge::Math,
    };

    settings.update_challenge(challenge)?;

    Ok(shellfirm::CmdExit {
        code: exitcode::OK,
        message: None,
    })
}
