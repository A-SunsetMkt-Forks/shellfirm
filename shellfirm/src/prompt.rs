use std::{io, thread, time::Duration};

use console::style;
use rand::Rng;

/// wrong answer text show when user solve the challenge incorrectly
const WRONG_ANSWER: &str = "wrong answer, try again...";
/// show math challenge text
const SOLVE_MATH_TEXT: &str = "Solve the challenge:";
/// show enter challenge text
const SOLVE_ENTER_TEXT: &str = "Type `Enter` to continue";
/// show yes challenge text
const SOLVE_YES_TEXT: &str = "Type `yes` to continue";
/// show yes challenge text
const DENIED_TEXT: &str = "The command is not allowed.";
/// show to the user how can he cancel the command
const CANCEL_PROMPT_TEXT: &str = "^C to cancel";

/// Show math challenge to the user.
pub fn math_challenge() -> bool {
    let mut rng = rand::thread_rng();
    let num_a = rng.gen_range(0..10);
    let num_b = rng.gen_range(0..10);
    let expected_answer = num_a + num_b;

    eprintln!(
        "{}: {} + {} = ? {}",
        SOLVE_MATH_TEXT,
        num_a,
        num_b,
        get_cancel_string()
    );
    loop {
        let answer = show_stdin_prompt();

        let answer: u32 = match answer.trim().parse() {
            Ok(num) => num,
            Err(_) => continue,
        };
        if answer == expected_answer {
            break;
        }
        eprintln!("{}", WRONG_ANSWER);
    }
    true
}

/// Show enter challenge to the user.
pub fn enter_challenge() -> bool {
    eprintln!("{} {}", SOLVE_ENTER_TEXT, get_cancel_string());
    loop {
        let answer = show_stdin_prompt();
        if answer == "\n" {
            break;
        }
        eprintln!("{}", WRONG_ANSWER);
    }
    true
}

/// Show yes challenge to the user.
pub fn yes_challenge() -> bool {
    eprintln!("{} {}", SOLVE_YES_TEXT, get_cancel_string());
    loop {
        if show_stdin_prompt().trim() == "yes" {
            break;
        }
        eprintln!("{}", WRONG_ANSWER);
    }
    true
}

/// Deny function will loop FOREVER until the user kill the process ^C.
/// it mean that the use command will never executed
pub fn deny() {
    eprintln!("{} type {}", DENIED_TEXT, get_cancel_string());
    loop {
        thread::sleep(Duration::from_secs(60));
    }
}

/// Catch user stdin. and return the user type
fn show_stdin_prompt() -> String {
    let mut answer = String::new();
    io::stdin()
        .read_line(&mut answer)
        .expect("Failed to read line");

    answer
}

/// return cancel string with colorize format
fn get_cancel_string() -> String {
    format!("{}", style(CANCEL_PROMPT_TEXT).underlined().bold().italic())
}
