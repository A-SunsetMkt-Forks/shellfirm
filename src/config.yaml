challenge: Math # Math, Enter, YesNo

checks:
  - is: git reset
    method: Contains
    enable: true
    description: "This command going to reset all your local changes"
  - is: "rm.+(-r|-f|-rf|-fr)*"
    method: Regex
    enable: true
    description: "You are going to deletes everything in the path"
  - is: ":(){ :|: & };:"
    method: Contains
    enable: true
    description: "This short line defines a shell function that creates new copies of itself.\nThe process continually replicates itself, and its copies continually replicate themselves, quickly taking up all your CPU time and memory.\nThis can cause your computer to freeze. Itâ€™s basically a denial-of-service attack."
  - is: ">.+/dev/sda"
    method: Regex
    enable: true
    description: "Writing the data directly to the hard disk drive and damaging your file system."
  - is: "mv+.*/dev/null"
    method: Regex
    enable: true
    description: "The files will be discarded and destroyed."