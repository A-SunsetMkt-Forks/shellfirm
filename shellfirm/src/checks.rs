//! Manage command checks

///
use crate::config::{Challenge, Method};
use crate::prompt;
use anyhow::Result;
use console::style;
use log::debug;
use rayon::prelude::*;
use regex::Regex;
use serde_derive::{Deserialize, Serialize};
use std::collections::HashMap;

// list of custom filter
#[derive(Debug, Deserialize, Serialize, PartialEq, Eq, Hash, Clone)]
pub enum FilterType {
    IsFileExists,
}

/// Describe single check
#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct Check {
    /// test ia a value that we check the command.
    pub test: String,
    /// The type of the check.
    pub method: Method,
    /// boolean for ignore check
    pub enable: bool,
    /// description of what is risky in this command
    pub description: String,
    /// the group of the check see files in `checks` folder
    pub from: String,
    #[serde(default)]
    pub challenge: Challenge,
    #[serde(default)]
    pub filters: HashMap<FilterType, String>,
}

/// prompt a challenge to the user
///
/// # Errors
///
/// Will return `Err` when could not convert checks to yaml
pub fn challenge(challenge: &Challenge, checks: &[Check], dryrun: bool) -> Result<bool> {
    if dryrun {
        eprintln!("{}", serde_yaml::to_string(checks)?);
        return Ok(true);
    }
    eprintln!("{}", style("#######################").yellow().bold());
    eprintln!("{}", style("# RISKY COMMAND FOUND #").yellow().bold());
    eprintln!("{}", style("#######################").yellow().bold());

    let mut descriptions: Vec<String> = Vec::new();
    for check in checks {
        if !descriptions.contains(&check.description) {
            descriptions.push(check.description.to_string());
        }
    }
    for description in descriptions {
        eprintln!("* {}", description);
    }
    eprintln!();

    let show_challenge = challenge;

    Ok(match show_challenge {
        Challenge::Math | Challenge::Default => prompt::math_challenge(),
        Challenge::Enter => prompt::enter_challenge(),
        Challenge::Yes => prompt::yes_challenge(),
    })
}

/// Check if the given command matched to on of the checks
///
/// # Arguments
///
/// * `checks` - List of checks that we want to validate.
/// * `command` - Command check.
#[must_use]
pub fn run_check_on_command(checks: &[Check], command: &str) -> Vec<Check> {
    checks
        .par_iter()
        .filter(|&v| v.enable)
        .filter(|&v| is_match(v, command))
        .filter(|&v| check_custom_filter(v, command))
        .map(std::clone::Clone::clone)
        .collect()
}

/// filter custom checks
///
/// When true is returned it mean the filter should keep the check and not filter our the check.
///
/// # Arguments
///
/// * `check` - Check struct
/// * `command` - Command.
fn check_custom_filter(check: &Check, command: &str) -> bool {
    // if check is not regex type or custom filters are not configure return true.
    if check.method != Method::Regex || check.filters.is_empty() {
        return true;
    }
    // Capture command groups from the current check
    let caps = Regex::new(&check.test).unwrap().captures(command).unwrap();

    // by default true is return. it mean the check not filter out (safe side security).
    let mut keep_check = true;
    for (filter_type, filter_params) in &check.filters {
        debug!(
            "filter information: command {} include filter: {:?} filter_params: {}",
            command, filter_type, filter_params
        );

        let keep_filter = match filter_type {
            FilterType::IsFileExists => filter_is_file_exists(
                caps.get(filter_params.parse().unwrap())
                    .map_or("", |m| m.as_str()),
            ),
        };

        if !keep_filter {
            keep_check = false;
            break;
        }
    }

    keep_check
}

/// Check if the given command match to one of the existing checks
///
/// # Arguments
///
/// * `check` - Check struct
/// * `command` - command for the check
fn is_match(check: &Check, command: &str) -> bool {
    match check.method {
        Method::Contains => is_contains(&check.test, command),
        Method::StartWith => is_start_with(&check.test, command),
        Method::Regex => is_regex(&check.test, command),
    }
}

/// Is the command contains the given check.
///
/// # Arguments
///
/// * `test` - check value
/// * `command` - command value
fn is_contains(test: &str, command: &str) -> bool {
    command.contains(test)
}

/// is the command start with the given check.
///
/// # Arguments
///
/// * `test` - check value
/// * `command` - command value
fn is_start_with(test: &str, command: &str) -> bool {
    command.starts_with(test)
}

/// Is the command match to the given regex.
///
/// # Arguments
///
/// * `test_r` - check value
/// * `command` - command value
fn is_regex(test_r: &str, command: &str) -> bool {
    Regex::new(test_r).unwrap().is_match(command)
}

/// check if the path exists (file and folder).
///
/// # Arguments
///
/// * `file_path` - check path.
fn filter_is_file_exists(file_path: &str) -> bool {
    let mut file_path: String = file_path.trim().into();
    if file_path.starts_with('~') {
        match dirs::home_dir() {
            Some(path) => {
                file_path = file_path.replace('~', &path.display().to_string());
            }
            None => return true,
        };
    }
    debug!("check is file {} exists", file_path);
    return std::path::Path::new(file_path.trim()).exists();
}

#[cfg(test)]
mod test_checks {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn is_match_command() {
        let regex_check = Check {
            test: String::from("rm.+(-r|-f|-rf|-fr)*"),
            method: Method::Regex,
            enable: true,
            description: String::from(""),
            from: String::from(""),
            challenge: Challenge::Default,
            filters: HashMap::new(),
        };
        let contains_check = Check {
            test: String::from("test"),
            method: Method::Contains,
            enable: true,
            description: String::from(""),
            from: String::from(""),
            challenge: Challenge::Default,
            filters: HashMap::new(),
        };
        let startwith_check = Check {
            test: String::from("start"),
            method: Method::StartWith,
            enable: true,
            description: String::from(""),
            from: String::from(""),
            challenge: Challenge::Default,
            filters: HashMap::new(),
        };
        assert_debug_snapshot!(&regex_check);
        assert_debug_snapshot!(&contains_check);
        assert_debug_snapshot!(&startwith_check);
    }
    #[test]
    fn can_check_is_contains() {
        assert_debug_snapshot!(is_contains("test", "test is valid"));
        assert_debug_snapshot!(is_contains("test is valid", "not-found"));
    }

    #[test]
    fn can_check_is_start_with() {
        assert_debug_snapshot!(is_start_with("test is", "test is valid"));
        assert_debug_snapshot!(is_start_with("test is valid", "is"));
    }

    #[test]
    fn can_check_is_regex_match() {
        assert_debug_snapshot!(is_regex("rm.+(-r|-f|-rf|-fr)*", "rm -rf"));
        assert_debug_snapshot!(is_regex("^f", "rm -rf"));
    }
}
