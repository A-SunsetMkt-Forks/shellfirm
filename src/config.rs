//! Configuration management

use crate::checks::Check;
use crate::cli::{UPDATE_CONFIGURATION_ONLY_DIFF, UPDATE_CONFIGURATION_OVERRIDE};
use anyhow::anyhow;
use anyhow::Result as AnyResult;
use serde_derive::{Deserialize, Serialize};
use std::fs;
use std::io::{Read, Write};

pub const DEFAULT_CONFIG_FILE: &str = include_str!("config.yaml");

/// The method type go the check.
#[derive(Debug, Deserialize, Serialize, Clone)]
pub enum Method {
    /// If the command start with.
    StartWith,
    /// if the command contains.
    Contains,
    /// if the command match to the given regex.
    Regex,
}

/// The user challenge when user need to confirm the command.
#[derive(Debug, Deserialize, Serialize)]
pub enum Challenge {
    /// Math challenge.
    Math,
    /// Only enter will approve the command.
    Enter,
    /// only yes/no typing will approve the command.
    YesNo,
}

/// describe configuration folder
pub struct SettingsConfig {
    /// Configuration folder path.
    pub path: String,
    /// config file.
    pub config_file_path: String,
    /// If configuration path overridden by the user.
    pub default: bool,
}

/// Describe the configuration yaml
#[derive(Debug, Deserialize, Serialize)]
pub struct Config {
    /// Type of the challenge.
    pub challenge: Challenge,
    /// List of checks.
    pub checks: Vec<Check>,
}

// /// Describe single check
// #[derive(Debug, Deserialize, Clone)]
// pub struct Check {
//     pub is: String,
//     pub method: Method,
//     pub enable: bool,
//     pub description: String,
// }

impl SettingsConfig {
    /// Convert config yaml file to struct.
    pub fn load_config_from_file(&self) -> AnyResult<Config> {
        Ok(serde_yaml::from_str(&self.read_config_file()?)?)
    }

    /// Return default app config as a config struct.
    pub fn load_default_config(&self) -> AnyResult<Config> {
        Ok(serde_yaml::from_str(DEFAULT_CONFIG_FILE)?)
    }

    /// Manage configuration folder/ file.
    /// * Create config folder if not exists
    /// * Create config yaml file if not exists
    pub fn manage_config_file(&self) -> AnyResult<()> {
        self.create_config_folder()?;
        if fs::metadata(&self.config_file_path).is_err() {
            self.create_default_config_file()?;
        }
        Ok(())
    }

    pub fn update_config_content(&self, behavior: &str) -> AnyResult<()> {
        match behavior {
            UPDATE_CONFIGURATION_ONLY_DIFF => self.add_diff_configuration_file(),
            UPDATE_CONFIGURATION_OVERRIDE => self.override_configuration_file(),
            _ => return Err(anyhow!("unsupported behavior")),
        }
    }

    /// Create config folder if not exists.
    fn create_config_folder(&self) -> AnyResult<()> {
        if let Err(err) = fs::create_dir(&self.path) {
            if err.kind() != std::io::ErrorKind::AlreadyExists {
                return Err(anyhow!("could not create folder: {}", err.to_string()));
            }
        }
        Ok(())
    }

    /// Create config file with default configuration content
    fn create_default_config_file(&self) -> AnyResult<()> {
        let mut file = fs::File::create(&self.config_file_path)?;
        file.write_all(DEFAULT_CONFIG_FILE.as_bytes())?;
        Ok(())
    }

    /// Create config file with default configuration content
    fn create_config_file_from_struct(&self, config: &Config) -> AnyResult<()> {
        let content = serde_yaml::to_string(config)?;
        let mut file = fs::File::create(&self.config_file_path)?;
        file.write_all(content.as_bytes())?;
        Ok(())
    }

    /// Convert config file to config struct.
    fn read_config_file(&self) -> AnyResult<String> {
        let mut file = std::fs::File::open(&self.config_file_path)?;
        let mut content = String::new();
        file.read_to_string(&mut content)?;
        Ok(content)
    }

    /// override existing config yaml file with the application configuration
    fn override_configuration_file(&self) -> AnyResult<()> {
        self.create_default_config_file()
    }

    /// adding only new check into configuration file.
    fn add_diff_configuration_file(&self) -> AnyResult<()> {
        match self.load_config_from_file() {
            Ok(mut conf) => {

                let default_app_config =  self.load_default_config()?;

                for default_check in default_app_config.checks.clone() {
                    let mut found = false;
                    for user_check in &conf.checks{
                        if user_check.is == default_check.is {found = true; break;}
                    }
                    if !found {
                        conf.checks.push(default_check.clone())
                    }
                }
                self.create_config_file_from_struct(&conf)?;
                Ok(())
            },
            Err(_e) => return Err(anyhow!("could not parse current config file. please try to fix the yaml file or override the current configuration by use the flag `--behavior override`"))
        }
    }
}

/// Get config config application details.
///
/// # Arguments
///
/// * `path` - Config folder path. if is empty default path will be returned.
pub fn get_config_folder(path: &str) -> AnyResult<SettingsConfig> {
    let package_name = env!("CARGO_PKG_NAME");

    let mut config_folder = path.into();
    let mut is_default = false;
    if path.is_empty() {
        match home::home_dir() {
            Some(path) => {
                is_default = true;
                config_folder = format!("{}/.{}", path.display(), package_name);
            }
            None => return Err(anyhow!("could not get directory path")),
        }
    }
    Ok(SettingsConfig {
        path: config_folder.clone(),
        default: is_default,
        config_file_path: format!("{}/config.yaml", config_folder),
    })
}

#[cfg(test)]
mod config {
    use super::*;

    fn get_current_project_path() -> String {
        std::env::current_dir().unwrap().to_str().unwrap().into()
    }
    #[test]
    fn can_get_config_default_folder() {
        let conf = get_config_folder("").unwrap();
        assert_ne!(conf.path, "");
        assert!(conf.default);
        assert_eq!(conf.config_file_path, format!("{}/config.yaml", conf.path));
    }

    #[test]
    fn can_get_config_custom_folder() {
        let folder_path = "/custom/folder/path";
        let conf = get_config_folder(folder_path).unwrap();
        assert_eq!(conf.path, folder_path);
        assert!(!conf.default);
        assert_eq!(conf.config_file_path, format!("{}/config.yaml", conf.path));
    }

    #[test]
    fn can_load_config_from_file() {
        let settings_config = SettingsConfig {
            path: get_current_project_path(),
            default: false,
            config_file_path: format!("{}/src/config.yaml", get_current_project_path()),
        };

        assert!(settings_config.load_config_from_file().is_ok())
    }

    #[test]
    fn cant_load_config_from_file() {
        let settings_config = SettingsConfig {
            path: String::from(""),
            default: false,
            config_file_path: String::from(""),
        };

        assert!(settings_config.load_config_from_file().is_err())
    }

    #[test]
    fn can_write_config_file() {
        let tmp_folder = format!("{}/tmp", get_current_project_path());
        let config_file_path = format!("{}/config.yaml", &tmp_folder);
        if fs::metadata(&config_file_path).is_ok() {
            fs::remove_file(&config_file_path).unwrap();
        }

        let settings_config = SettingsConfig {
            path: format!("{}", tmp_folder),
            default: false,
            config_file_path: config_file_path,
        };

        assert!(settings_config.manage_config_file().is_ok());
        assert!(settings_config.read_config_file().is_ok());
    }

    #[test]
    fn can_load_default_config() {
        let conf = get_config_folder("").unwrap();
        assert!(conf.load_default_config().is_ok())
    }

    #[test]
    fn can_override_existing_config() {
        let settings_config = SettingsConfig {
            path: get_current_project_path(),
            default: false,
            config_file_path: format!("{}/tmp/override.yaml", get_current_project_path()),
        };

        // check if we file created successfully
        assert!(fs::File::create(&settings_config.config_file_path)
            .unwrap()
            .write_all("".as_bytes())
            .is_ok());

        // then read the file and make sure that the content is empty
        let file_content = settings_config
            .read_config_file()
            .unwrap_or(format!("error"));

        assert_eq!(file_content, "");

        // create the default configuration
        assert!(settings_config.override_configuration_file().is_ok());
        // make sure that the file is not empty
        assert!(!settings_config.read_config_file().unwrap().is_empty());
    }

    #[test]
    fn can_add_diff_configuration_file() {
        let settings_config = SettingsConfig {
            path: get_current_project_path(),
            default: false,
            config_file_path: format!("{}/tmp/add-diff.yaml", get_current_project_path()),
        };

        let orig_config = settings_config.load_default_config().unwrap();
        let mut config = settings_config.load_default_config().unwrap();

        // creates configuration file with only 1 check (we want to check that we not change existing check and just append new ones)
        config.checks = vec![Check {
            is: String::from("is value"),
            method: Method::Contains,
            enable: true,
            description: String::from("description"),
        }];

        // create configuration file with 1 check
        assert!(settings_config
            .create_config_file_from_struct(&config)
            .is_ok());

        // make sure that the configuration created with 1 check
        assert_eq!(
            settings_config
                .load_config_from_file()
                .unwrap()
                .checks
                .len(),
            config.checks.len()
        );

        // create the config diff command
        assert!(settings_config.add_diff_configuration_file().is_ok());

        // make sure that the count of check that change equal to existing config + the default application config
        assert_eq!(
            settings_config
                .load_config_from_file()
                .unwrap()
                .checks
                .len(),
            config.checks.len() + orig_config.checks.len()
        );
    }
}
