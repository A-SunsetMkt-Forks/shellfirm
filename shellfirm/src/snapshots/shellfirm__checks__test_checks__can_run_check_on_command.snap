---
source: shellfirm/src/checks.rs
expression: "run_check_on_command(&checks, \"test-1\")"
---
[
    Check {
        id: "",
        test: "test-(1)",
        method: Regex,
        description: "",
        from: "test-1",
        challenge: Default,
        filters: {},
    },
    Check {
        id: "",
        test: "test-(1|2)",
        method: Regex,
        description: "",
        from: "test-2",
        challenge: Default,
        filters: {},
    },
]
